manager/middlewares/kessel.go

```diff
@@ -0,0 +1,153 @@
+		},
+	}
+
+	switch c.Request.Method {
+	case "GET", "POST":
+		req.Relation = patchReadPerm
+	case "PUT", "DELETE":
+		req.Relation = patchWritePerm
+	}
+
+	return req, nil
+}
+
+func receiveAll(stream ListObjectStreamingClient) ([]*kesselAPIv2.StreamedListObjectsResponse, error) {
+	responses := make([]*kesselAPIv2.StreamedListObjectsResponse, 0)
+	for res, err := stream.Recv(); err != io.EOF; res, err = stream.Recv() {
+		if err != nil {
+			return nil, err
+		}
+		responses = append(responses, res)
+	}
+	return responses, nil
+}
+
+func processWorkspaces(workspaces []*kesselAPIv2.StreamedListObjectsResponse) (map[string]string, error) {
+	inventoryGroups := make(map[string]string, 2)
+	inventoryGroups[utils.KeyUngrouped] = "[]"
+
+	groups := make([]string, 0, len(workspaces))
+	for _, workspace := range workspaces {
+		group, err := utils.ParseInventoryGroup(&workspace.Object.ResourceId, nil)
+		if err != nil {
+			// couldn't marshal inventory group to json
+			continue
+		}
+		groups = append(groups, group)
+	}
+
+	if len(groups) == 0 {
+		return nil, errors.New("no workspaces were found")
+	}
+
+	inventoryGroups[utils.KeyGrouped] = fmt.Sprintf("{%s}", strings.Join(groups, ","))
+	return inventoryGroups, nil
+}
+
+func hasPermissionKessel(c *gin.Context) {
+	client, err := setupClient()
+	if err != nil {
+		utils.LogError("err", err.Error(), "failed to setup a Kessel service client")
+		c.AbortWithStatusJSON(http.StatusInternalServerError, utils.ErrorResponse{
+			Error: "Unexpecred server error", // missing cert or failed to make a new gRPC client
+		})
+		return
+	}
+
+	req, err := buildRequest(c)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusUnauthorized, utils.ErrorResponse{Error: "Invalid x-rh-identity header"})
+		return
+	}
+
+	// TODO: how long should the timeout be
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	stream, err := client.KesselInventoryService.StreamedListObjects(ctx, req)
+	if err != nil {
+		utils.LogError("err", err.Error(), "failed to establish a gRPC stream with Kessel")
+		c.AbortWithStatusJSON(http.StatusInternalServerError, utils.ErrorResponse{
+			Error: "Failed to communicate with Kessel RBAC service",
+		})
+		return
+	}
+
+	workspaces, err := receiveAll(stream)
+	if err != nil {
+		utils.LogError("err", err.Error(), "failed to receive from Kessel")
+		c.AbortWithStatusJSON(http.StatusInternalServerError, utils.ErrorResponse{
+			Error: "Failed to communicate with Kessel RBAC service",
+		})
+		return
+	}
+
+	inventoryGroups, err := processWorkspaces(workspaces)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusUnauthorized, utils.ErrorResponse{
+			Error: "You don't have access to this application",
+		})
+		return
+	}
+
+	c.Set(utils.KeyInventoryGroups, inventoryGroups)
+	utils.LogDebug("Kessel check successful")
+}
+
+func Kessel() gin.HandlerFunc {
+	if !config.KesselEnabled {
+		return func(_ *gin.Context) {}
+	}
+	return hasPermissionKessel
+}
```

Comment:
I think more readable way would be to skip Kessel middleware completely one level above.
 
```suggestion
func Kessel() gin.HandlerFunc {
  	return hasPermissionKessel
}
```
and in [routes.go](https://github.com/RedHatInsights/patchman-engine/pull/1708/files#diff-25efcda7d5d9757ce756465943f6cbbec9b5742bd5f9beda91a5d4bfc243db85)
```
if config.KesselEnabled {
  userAuth.Use(middlewares.Kessel())
}	
```